lib/ansible/plugins/action/reboot.py

```diff
@@ -67,53 +65,104 @@ def __init__(self, *args, **kwargs):
-        if pre_reboot_delay < 0:
-            pre_reboot_delay = 0
+    @property
+    def post_reboot_delay(self):
+        return self._check_delay('post_reboot_delay')
 
+    def _check_delay(self, key):
+        value = int(self._task.args.get(key, self._task.args.get(key + '_sec', self.DEFAULT_POST_REBOOT_DELAY)))
+        if value < 0:
+            value = 0
+        else:
+            return value
+
+    def _get_value_from_facts(self, variable_name, distribution, default_value):
+        ''' Get dist+version specific args first, then distribution, then family, lastly use default '''
+        attr = getattr(self, variable_name)
+        value = attr.get(
+            distribution['name'] + distribution['version'],
+            attr.get(
+                distribution['name'],
+                attr.get(
+                    distribution['family'],
+                    getattr(self, default_value))))
+        return value
+
+    def get_shutdown_command_args(self, distribution):
+        args = self._get_value_from_facts('SHUTDOWN_COMMAND_ARGS', distribution, 'DEFAULT_SHUTDOWN_COMMAND_ARGS')
         # Convert seconds to minutes. If less that 60, set it to 0.
-        delay_min = pre_reboot_delay // 60
-        msg = self._task.args.get('msg', self.DEFAULT_REBOOT_MESSAGE)
-
-        shutdown_command_args = shutdown_command_args.format(delay_sec=pre_reboot_delay, delay_min=delay_min, message=msg)
-
-        reboot_command = '%s %s' % (shutdown_command, shutdown_command_args)
-        return reboot_command
-
-    def get_system_boot_time(self):
-        stdout = u''
-        stderr = u''
-
-        # Determine the system distribution in order to use the correct shutdown command arguments
-        uname_result = self._low_level_execute_command('uname')
-        distribution = uname_result['stdout'].strip().lower()
+        delay_min = self.pre_reboot_delay // 60
+        reboot_message = self._task.args.get('msg', self.DEFAULT_REBOOT_MESSAGE)
+        return args.format(delay_sec=self.pre_reboot_delay, delay_min=delay_min, message=reboot_message)
+
+    def get_distribution(self, task_vars):
+        distribution = {}
+        display.debug('{action}: running setup module to get distribution'.format(action=self._task.action))
+        module_output = self._execute_module(
+            task_vars=task_vars,
+            module_name='setup',
+            module_args=dict(gather_subset='min'))
+        try:
+            if 'failed' in module_output.keys():
+                raise AnsibleError('Failed to determine system distribution. {0}, {1}'.format(
+                    to_native(module_output['module_stdout']).strip(),
+                    to_native(module_output['module_stderr']).strip()))
+            distribution['name'] = module_output['ansible_facts']['ansible_distribution'].lower()
+            distribution['version'] = str(module_output['ansible_facts']['ansible_distribution_version'].split('.')[0])
```

Comment:
Should you be using `to_text` here and the other places you are using `str()`?

Reply:
Yes! Bad habit.